File: src/com/aopphp/go/pointcut/InheritanceClassFilter.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.PhpClassHierarchyUtils;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;

import java.util.Collection;
import java.util.Collections;
import java.util.Set;

/**
 * Inheritance class matcher that match single class name or any subclass
 */
public class InheritanceClassFilter implements PointFilter {

    /**
     * Filter type
     */
    private static final Set<KindFilter> KIND_CLASS = Collections.singleton(KindFilter.KIND_CLASS);

    /**
     * Instance class name to match
     */
    private String parentClassName;

    public InheritanceClassFilter(String parentClassName) {

        this.parentClassName = parentClassName;
    }

    @Override
    public Set<KindFilter> getKind() {
        return KIND_CLASS;
    }

    @Override
    public boolean matches(PhpNamedElement element) {
        if (!(element instanceof PhpClass)) {
            return false;
        }

        Collection<PhpClass> parentClasses = PhpIndex.getInstance(element.getProject()).getAnyByFQN(parentClassName);
        if (parentClasses.size() == 0) {
            return false;
        }

        PhpClass parentClass = parentClasses.iterator().next();

        return PhpClassHierarchyUtils.getAllSubclasses(parentClass).contains(element);
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/InheritanceClassFilter.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpClass
import com.jetbrains.php.lang.psi.elements.PhpNamedElement

class InheritanceClassFilter(private val parentClassName: String) : PointFilter {
    private val _kind = setOf(KindFilter.KIND_CLASS)
    private val normalizedFqn = if (parentClassName.startsWith("\\")) parentClassName else "\\$parentClassName"

    override fun getKind() = _kind

    override fun matches(element: PhpNamedElement): Boolean {
        if (element !is PhpClass) return false
        val elementFqn = element.fqn ?: return false
        if (elementFqn == normalizedFqn) return true
        return element.isSubclassOf(normalizedFqn)
    }

    private fun PhpClass.isSubclassOf(fqn: String): Boolean {
        var current: PhpClass? = superClass
        while (current != null) {
            if (current.fqn == fqn) return true
            current = current.superClass
        }
        return implementedInterfaces.any { it.fqn == fqn }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is InheritanceClassFilter) return false
        return parentClassName == other.parentClassName
    }

    override fun hashCode() = parentClassName.hashCode()
}
